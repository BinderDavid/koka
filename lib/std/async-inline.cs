/*---------------------------------------------------------------------------
  Copyright 2017 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/
using System.Threading;

static class _Async
{
  public class ThreadTimer {
    private Primitive.EventloopEntry entry;
    private Timer timer;

    public ThreadTimer( Fun0<Unit> cb, int ms ) {
      entry = Primitive.GetEventloopEntry();
      if (ms <= 0) {
        timer = null;
        entry.Post(() => { cb.Apply(); });
      }
      else {
        timer = new Timer( (object state0) => { 
                  if (entry != null) entry.Post(() => { cb.Apply(); }); 
                }, null, ms, Timeout.Infinite );
      }
    }

    public void Close() {
      if (entry != null) entry.Close();
      if (timer != null) timer.Dispose();
      entry = null;
      timer = null;
    }
  }

  public static ThreadTimer SetTimeout( Fun0<Unit> cb, int ms ) {
    return new ThreadTimer(cb,ms);
  }

  public static void ClearTimeout( object obj ) {
    if (obj == null) return;
    ThreadTimer t = (ThreadTimer)obj;
    t.Close();
  }
}
