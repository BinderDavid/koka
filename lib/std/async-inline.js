/*---------------------------------------------------------------------------
  Copyright 2012-2016 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

var effCancel = "std/async/cancelable";

function _create_cancelable( caction ) {
  // identity operations
  function ret( loc, x, _k ) { return _k(x); };
  function ops(_l1, _l2, _cont, _op, _resume , _k ) { return _resume(_l1, _l2, _cont, _unit_, _k);  };
  var handle = $std_core._makeHandler1( effCancel, ret, ops );
  var loc = { canceled:false, cancelers:[] };
  function action(_k) { return caction( function(){ return _cancel(loc); }, _k); }
  return function(_k){ return handle( loc, action, _k ); } 
}

function throw_canceled() {
  $std_core.throw_1("Canceled",Cancel);
}

function _cancel( loc ) {
  loc.canceled = true;
  var cs = loc.cancelers; loc.cancelers = [];
  cs.forEach( function(cb) { 
    if (typeof cb.cancel === "function") { 
      cb.cancel();
      setTimeout( function() {
        $std_core._restore_handler_context(cb.hctx, function(_) {
          throw_canceled(); // throw a Cancel exception in that context
        });
      }, 0);
    }
  });
  //return $std_core._unit_;
}

function _check_canceled( hctx ) {
  if (hctx==null) hctx = $std_core._save_handler_context();
  var top = hctx.htop;
  while (top >= 0) {
    var h = hctx.hstack[top];
    if (h.optag===effCancel && h.loc1.canceled) throw_canceled();
    top--;
  }
  //return $std_core._unit_;
}

function _register_cancel( cb, f ) {
  var hctx = $std_core._save_handler_context();
  var top = hctx.htop;
  while (top >= 0 && hctx.hstack[top].optag !== effCancel) { top--; }
  if (top >= 0) {
    //console.log("set cancel: " + cb + ", f: " + f);
    // only register if a cancelation frame exists
    cb.loc = hctx.hstack[top].loc1
    cb.loc.cancelers.push(cb);
    cb.cancel = f;
  }
  //return $std_core._unit_;
}

function _unregister_cancel( cb ) {
  if (typeof cb.cancel === "function") {
    cb.loc.cancelers = cb.loc.cancelers.filter( function(f) { return (f !== cb); } );
  }
  //return $std_core._unit_;
}



// --------------------------------------------------------
// Support for asynchronous actions with effect handlers
// --------------------------------------------------------

var _syncNesting = 0;
var _syncBlocked = null;

function $asyncStartSynchronous() {
  _syncNesting++;
  if (_syncBlocked===null) _syncBlocked = [];
}

function $asyncEndSynchronous() {
  if (_syncNesting===0) return;
  _syncNesting--; 
  if (_syncNesting===0) {
    // now invoke all queued callbacks; we schedule all through a timeout
    _syncBlocked.forEach( function(entry) {
      var cb = function() { return entry.fun.apply(entry.fun,entry.args); };
      setTimeout(cb,0);
    });
    _syncBlocked = null;
  }
}

// Create a callback wrapper around 'cb'. 
// If 'haserr' is true, the first parameter of the new call back is considered an error object
// and thrown automatically if not null. If it is null, it is not passed to the callback `cb`.
// If `nodup` is true, multiple call backs that are queued due to synchronized blocks are not duplicated
// and only the last call back is queued (for timers, mouse events etc).
function $asyncCallback( cb, haserr, nodup ) {
  // save the handler context
  var _hctx = $std_core._save_handler_context();
  // check for cancelation
  _check_canceled(_hctx);
  // remember if this callback itself was constructed inside a sync block
  var _in_synchronous = (_syncNesting > 0); 
  // and return a wrapped callback that restores the context
  var self = null;
  self = (function() {
    var args = Array.prototype.slice.call(arguments);
    // unregister cancelation
    _unregister_cancel(self);
    // check if we return while a 'synchronous' block was started
    if (!self.in_synchronous && _syncNesting > 0) { // if so, put ourselves on the queue
      if (nodup) {
        // update in-place possible previous occurrence of this event if 'nodup' was set (for timers or mouse events etc.)
        for (var i = 0; i < _syncBlocked.length; i++) {
          if (_syncBlocked[i].fun == self) {
            _syncBlocked[i].args = args;
            return; // updated in place, we are done.
          }
        }        
      }
      // and push ourselves on the queue to be called later
      _syncBlocked.push( { fun: self, args: args } );
      return;
    }
    else {
      // restore the handler context and invoke the callback
      $std_core._restore_handler_context( self.hctx, function(_) {
        if (haserr) {
          if (args[0] != null) throw args[0];
          args.shift();         
        }
        return cb.apply(cb,args);
      }, $std_core._unit_ );
    }
  });
  self.hctx = _hctx;
  self.in_synchronous = _in_synchronous;
  self.loc = null;
  self.cancel = null;
  self.canceled = false;
  return self;
}

function $asyncFork( action ) {
  setTimeout( _asyncCallback( function() {
    action($std_core.async_yield);
  }), 0 );  
}

function $asyncAll( actions, k ) {
  var n       = actions.length;
  var results = new Array(n);
  var exn;
  if (n<=0) return k(results);
  
  // all concurrent actions end with this continuation
  // only when all are finished, we resume the passed in continuation
  function allk(i,x,f) {
    results[i] = x;
    n--;
    if (n <= 0) {
      // if an exception was set, raise it now (or keep raising on multiple resume!)
      if (exn!==undefined) throw exn;
      // in case a branch exits more than once (due to amb for example)
      // note: this should never happen as multiple resumes can only happen after
      // at least one resumption but that cannot be past an asyncAll..
      for(i = 0; i < results.length; i++) {
        if (results[i] === undefined) return f(x); 
      }
      return k(results);
    }
    else {
      return f(x);
    }
  }

  for( var i = 0; i < n; i++) {
    (function(j){ 
      var cb = $asyncCallback( function(_) {
        return $std_core.$catch( function(k1) {
            return actions[j](k1);
          },
          function(_exn,_k2) {
            if (exn===undefined) exn = _exn;
            return allk(j, undefined, $std_core.async_yield);
          },
          function(x) { 
            return allk(j,x,$std_core.Result); 
          });
        }); 
      // by using setTimeout we can schedule all continuations for execution
      setTimeout(cb,0); 
    })(i);  // capture i by value
  }
  return $std_core.async_yield();
}
