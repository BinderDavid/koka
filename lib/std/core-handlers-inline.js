/*---------------------------------------------------------------------------
  Copyright 2017 Daan Leijen, Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

/*---------------------------------------------------------------------------
General notes:

- _funtions starting with an underscore are meant to be private

General design:

* Bind and Yield:
  - To get results from effectful functions use `bind`. For example, `return bind( get(), function(i) {... })`
  - Effectful functions either return with their result, or they _yield_ an operation.
  - A yielded operation is an object of the form:
    ```
    { _yield    : bool    // special field to recognize yielding an operation versus a regular return 
      cont      : a -> b  // the continuation to resume with a result for the yielded operation (like `moveNext`)
      op_arg    : c       // the operation argument (like an integer for a `set` operation)
      handler   : handler // the handler that handles the effect
      branch    : branch  // the branch in the handler that handles this operation
    }
    ```
    where the operation has type, `c -> a`, and the action argument to the handler has a
    result type `b`.
  - A bind operation extends the continuation field `cont` to be able to resume correctly

* Handlers and the handler stack:
  - A handler is described as:
    ```
    { effect_tag: string    // the effect being handled, like `amb`
      return_fun: function  // a function applied to the result of the action passed to the handler
      branches  : [branch]  // a vector of branches, one for each operation in the effect
    }
    ```
  - Handlers are pushed on the global handler stack when enabled
  - When an operation yields, it first finds the innermost handler on this handler stack 
    and also the branch that handles the particular operation.

* Branches
  - A branch handles a particular operation:
    ```
    { op_tag     : string,   // the operation tag being handled, like `get` or `set`
      resume_kind: int,      // the kind of resume: a tail-resumptive branch or a regular one? 
      branch_fun : (resume,op_arg,local) -> a   // the actual handling function for `op_tag`
    }
    ```

* Tail resumptive branches:
  - For branches that are _tail resumptive_, ie. do a `resume` as their final tail call, we can 
    now optimize by calling the branch directly as direct function call: there is no need to
    yield explicity and unwind the stack and finally resume again.
  - But when optimizing such tail-resumptive call, we need to be careful that any operations called
    _inside_ the handler are only handled by handlers above it. For this to work we use _skip frames_.

* Tail-recursive effectful operations
  - Finally, for tail-recursive functions that use `bind`, we have a special `tailcall` yield that
    unwinds to the nearest bind and then initiates the tail call such that the stack space stays 
    constant. See `tailcall` and the `_tailcalling`.

* Skip frames
  - todo

* Exceptions
  - todo

---------------------------------------------------------------------------*/

const Eff = (function() {
  // The identity function
  function id(x) {
    return x;
  }

  // For assertions
  function _assert(b,msg) {
    if (!b) {
      msg = msg || "no message";
      throw new Error("Assertion failed: " + msg );
    }
  }


  // Different resume kinds for branches: used for potential optimizations (`Normal` is always correct to use).
  // Especially branches that `Tail` resume are very common and should be optimized well: in this library
  // these are resolved to direct method calls without needing to unwind/rewind the stack.
  const _resume_kind = {
    Default : 0,
    Never   : 1,    // never calls `resume`.
    Tail    : 2,    // calls `resume` at most once as the last thing to do (or does not resume at all).
    Once    : 3,    // calls `resume` at most once.
    Normal  : 4     // first-class resumption: calls `resume` once or many times.
  };


  /*----------------------------------------------------------------
    Bind, yield, and tailcall
  ----------------------------------------------------------------*/

  // Is this is a yielding value? Relies on the `_yield` field being special.
  function _yielding(x) {
    return (x != null && x._yield === true);
  }

  // Is this a tailcalling function? Optimization for effectful recursive functions to not use up the stack.
  function _tailcalling(x) {
    return (x != null && x._tailcall != null);
  }

  // Handle tailcall yields recursively
  function _handle_tailcalls(x) {
    while(_tailcalling(x)) { x = x._tailcall(); }
    return x;
  }

  // Bind a result of an effectful computation to the `next` function. 
  // When the effectful computation yielded on operation, we can remember where to continue (namely the `next` function)
  function bind( x, next ) 
  {
    x = _handle_tailcalls(x);
    if(_yielding(x)) {
      // an operation was yielded, re-yield with an updated resume continuation
      // note: we could perhaps maintain an array of `bind next` 
      //       functions instead of allocating separate closures.
      if (x.branch.resume_kind !== _resume_kind.Never) {
        const cont = x.cont; // capture locally
        x.cont = function(arg) { 
          return bind( cont(arg), next ); 
        };
      }
      // and return the updated yield value
      return x;
    }
    else {
      // not yielding, pass on the value to the next thing to do.
      return next(x);
    }
  }

  // Yield an operation consists of yielding a return object.
  function yield_op( effect_tag, op_tag, op_arg ) {
    // find our handler in the global handler stack
    const r = _find_handler(effect_tag,op_tag);
    if (r.branch.resume_kind !== _resume_kind.Tail) {
      // The handler uses a regular resumption, start unwinding the through the binds upto the handler
      return { 
        _yield    : true,               // signify we are yielding
        op_arg    : op_arg,             // the operation arguments (the new state)
        cont      : id,                 // the resume continuation: called with the operation result,
        handler   : r.handler,          // to handler to where to yield
        branch    : r.branch,           // the handler branch that matches our operation      
      }
    }
    else {
      // The handler uses a tail resumption: optimize it by directly 
      // calling the branch function without any unwinding.
      // This is important in practice as 90% of all operation handlers are tail-resumptions.
      const skiphandler = { effect_tag: _effect_skip, skip: r.skip, handler: r.handler };  // todo: avoid this allocation with special skip frames?
      return _skip_handle(skiphandler, r.branch.branch_fun, op_arg );
    }
  }

  // Yield a tailcall; use this for tail-recursive effectful computations to not use unneeded stack.
  function tailcall( action ) {
    return {
      _yield   : true,
      _tailcall: action
    };
  }


  /*----------------------------------------------------------------
    The handler stack
  ----------------------------------------------------------------*/

  // The global handler stack
  const handlers = [];

  function _handlers_push(h) {
    handlers.push(h);
  }

  function _handlers_pop(expected) {
    const h = handlers.pop();
    _assert(h === expected, "Handler stack is out of sync!");
  }

  // Find the handler for a particular effect and operation.
  function _find_handler( effect_tag, op_tag ) {
    // find the handler that handles the given effect
    var skipping = 0;
    for(var i = handlers.length - 1; i >= 0; i--) {
      let h = handlers[i];      
      if (skipping > 0) {
        // skip when needed due to a skip frame
        skipping--;
      }
      else if (h.effect_tag == effect_tag) {
        // we found a handler for our effect
        for(var j = 0; j < h.branches.length; j++) {
          // find a matching branch for the operation
          let b = h.branches[j];
          if (b.op_tag == op_tag) {
            return { handler: h, branch: b, skip: handlers.length - i };
          }
        }
        throw new Error("Bad handler: Operation " + op_tag + " is not found in the handler for " + effect_tag);
      } 
      else if (h.effect_tag == _effect_skip) {
        // on skip frames, we start skipping
        skipping = h.skip;
      }
    }
    throw new Error("There is no handler for " + effect_tag + "/" + op_tag);
  }



  /* ----------------------------------------------------------------
    Skip frames: skip on a first read :-)

    When doing direct calls to tail-resumptive branches, we push
    a skip frame to ensure that any operations in a handler are not handled
    by any operations under it in the handler stack.

    There are other approaches, like popping the handler stack,
    but this seems most efficient?
  ----------------------------------------------------------------*/

  // Special effect tag of skip handlers
  const _effect_skip = "<skip>";

  // The tail resume just returns the results in a record.
  function _tail_resume(x,local) {
    return { _result: x, local: local };  // we assume `_result` is a special field
  }

  // Call a tail-resumptive branch using a skip frame handler
  function _skip_handle( skiphandler, branch_fun, op_arg) {
    _handlers_push(skiphandler);
    const x = _handle_tailcalls( branch_fun(_tail_resume, op_arg, skiphandler.handler.local) );  
    _handlers_pop(skiphandler);
    return _skip_handle_yield( skiphandler, x );
  }

  // Resume in a tail-resumptive branch 
  function _skip_resume( skiphandler, cont, arg ) {
    _handlers_push(skiphandler);
    const x = _handle_tailcalls( cont(arg) );
    _handlers_pop(skiphandler);
    return _skip_handle_yield( skiphandler, x );
  }

  // Handle the result of a tail-resumptive branch 
  function _skip_handle_yield( skiphandler, x ) {
    if (_yielding(x)) {
      // extend the continuation with the skip frame 
      _assert(!_tailcalling(x), "_skip_handle_yield: tail calls should already have been handled.");
      if (x.branch.resume_kind !== _resume_kind.Never) {
        const cont = x.cont; // capture locally
        x.cont = function(arg) {
          return _skip_resume( skiphandler, cont, arg );
        }
      }
      // and reyield
      return x;
    }
    else if (x != null && x._result !== undefined) {
      // we tail-resumed
      // record the new local state of the handler in-place
      if (x.local !== undefined) {
        skiphandler.handler.local = x.local;
      }
      return x._result;
    }
    else {
      // we returned normally without resuming!
      // in this case we need to unwind back to the handler anyway and return from there.
      // yield specially with a `_tail_return_branch` to signify this.
      return {
        _yield: true,
        op_arg: x,
        cont: id,
        handler: skiphandler.handler,
        branch: _tail_return_branch
      };
    }
  }

  // The special tail return branch is a default branch that always returns its argument directly.
  const _tail_return_branch = {
    op_tag: "<tail return>",
    resume_kind: _resume_kind.Never,
    branch_fun: function(resume,op_arg,local) {
      return op_arg;
    }
  };


  /*----------------------------------------------------------------
    Create an operation branch; i.e. one match in a handler
  ----------------------------------------------------------------*/

  // Create an handler branch, i.e. the function that is called for a particular operation
  function new_branch( resume_kind, op_tag, branch_fun ) {
    return {
      op_tag: op_tag,
      resume_kind: resume_kind || _resume_kind.Normal,
      branch_fun: branch_fun,
    };
  }

  function new_branch_never( op_tag, branch_fun ) {
    return new_branch(_resume_kind.Never,op_tag,branch_fun);
  }

  function new_branch_once( op_tag, branch_fun ) {
    return new_branch(_resume_kind.Once,op_tag,branch_fun);
  }

  function new_branch_tail( op_tag, branch_fun ) {
    return new_branch(_resume_kind.Tail,op_tag,branch_fun);
  }

  function new_branch_normal( op_tag, branch_fun ) {
    return new_branch(_resume_kind.Normal,op_tag,branch_fun);
  }



  /*----------------------------------------------------------------
    Create a handler
  ----------------------------------------------------------------*/

  // Create a handler for a given effect
  function new_handler( effect_tag, return_fun, branches ) 
  {
    branches = branches || [];
    return_fun = return_fun || id;

    // shared handler if there is no local state
    const shared_handler = {
      effect_tag: effect_tag,
      branches: branches
    };
    
    // handle an action or resumption
    function handle(action,local,arg) {
      // only share handler record if no local state is needed; otherwise we need to recreate the handler
      // to ensure with tail call optimizations the local state is not shared among different handler instantiations.
      const h = (local === undefined ? shared_handler : { effect_tag: effect_tag, branches: branches, local: local });
      _handlers_push(h);
      const x = _handle_tailcalls( action(arg) );
      _handlers_pop(h);
      return handle_yield(x, h, h.local);
    }

    // process the result of an action or resumption
    function handle_yield(x, h, local) {
      if (_yielding(x)) {
        _assert(!_tailcalling(x), "handle_yield: tail calls should already have been handled");
        // are we yielding to this handler?
        if (x.handler !== h) {
          // no, update the continuation
          if (x.branch.resume_kind !== _resume_kind.Never) {
            const cont = x.cont;         // capture local continuation
            x.cont = function(arg) { 
              return handle(cont,local,arg); 
            };          
          }
          // and re-yield
          return x;
        }
        else {
          h = null;           // for gc
          x.handler = null;   // for gc
          // handle the operation
          // note: no need to optimize tail resumptions, already handled in `yield`
          if (x.branch.resume_kind !== _resume_kind.Never) {
            const cont = x.cont;                  // capture local continuation
            function resume(result,newlocal) {    // create a resuming function
              return handle(cont, (newlocal!==undefined ? newlocal : local), result);  
            }
            return x.branch.branch_fun( resume, x.op_arg, local );          
          }
          else {
            return x.branch.branch_fun( _never_resume, x.op_arg, local ); 
          }       
        }
      }
      else {
        // a regular result without yielding
        return return_fun(x, local);
      }
    }

    // return a handler function
    return (function(action,local) { 
      return handle(action,local,undefined); 
    });
  }

  function new_handler1( effect_tag, return_fun, branches ) {
    function return_fun_swap(result,local) {
      return return_fun(local,result);
    }
    const f = new_handler(effect_tag,return_fun_swap,branches);
    return (function(local,action) { return f(action,local); });
  }

  function new_branch1( resume_kind, op_tag, branch_fun ) {
    function branch_fun_swap( resume, op_arg, local ) {
      return branch_fun( function(newlocal,result) { 
        return resume(result,newlocal); 
      }, op_arg, local );
    }
    return new_branch(resume_kind,op_tag,branch_fun_swap);
  }

  function _never_resume() {
    throw new Error("A handler branch marked as never resuming tried to resume.")
  }


  /*----------------------------------------------------------------
    Extension: functions to work with generator style yielding of operations.
  ----------------------------------------------------------------*/

  // Transform the result of an effectful iterator to regular operation yields.
  function yield_iter( x ) {
    return (_is_iterator(x) ? _yield_iter_resume(x,undefined) : x);
  }

  // Is this an iterator?
  function _is_iterator(x) {
    return (x != null && typeof x.next === "function");
  }

  // Handle operation yields that are yielded through an iterator.
  function _yield_iter_resume( iter, arg ) {
    let res = iter.next(arg);
    const x = res.value;
    if (!res.done) {
      _assert(_yielding(x), "regular operations should be wrapped in `to_iter` to lift them to an iterator (function*)");
      _assert(x.branch.resume_kind < _resume_kind.Normal, "effects yielded in iterators should resume at most once!");
      // _assert(x.cont === id, "effects yielded inside an iterator should not be bound using `bind`.");
      const cont = x.cont;  // capture locally 
      x.cont = function(arg) {
        return bind(cont(arg), function(argval) {  // bind in case of yielding over a regular yielding computation
          return _yield_iter_resume(iter,argval);
        });
      };
    }
    return x;
  }

  // Use this instead of a `yield` keyword to optimize for tail-resumptive operations that might
  // return a value right away.
  // Note: perhaps we should implement a `next` method on the yield object so we can use the `yield*` keyword?
  function* to_iter( x ) {
    if (_yielding(x)) {
      return yield x;
    }
    else {
      return x;
    }
  }

  /*----------------------------------------------------------------
    Extension: make it possible to catch exceptions over effectful 
    computations; in particular, this scopes over the `next` continuations
    of the `bind`s.
  ----------------------------------------------------------------*/

  // An exception handler. 
  // There is no need to push on the handler stack since there is 
  // no associated operation (i.e. it is handled by the runtime system already)
  function handle_exn( action, on_exn ) {
    return _handle_exn_resume( action, on_exn, undefined);
  }

  function _handle_exn_resume( action, on_exn, arg) {
    const n = handlers.length; // remember the handler stack length so we can restore it on an exception
    try {
      const x = _handle_tailcalls(action(arg));
      _assert(handlers.length === n, "Handler stack is out of sync in a try block");
      if (_yielding(x)) {
        let cont = x.cont; // capture locally
        x.cont = function(arg) {
          return _handle_exn_resume(cont, on_exn, arg);  // again execute within the try block
        };
      }   
      return x;
    }
    catch(exn) {
      _assert(handlers.length >= n, "After exception the handlers stack is too shallow");
      if (handlers.length > n) {
        handlers.length = n; // restore handler stack 
      }
      return on_exn(exn);
    }
  }

  let _in_bind_context = false;
  function is_in_bind_context() {
    return _in_bind_context;
  }

  function set_in_bind_context(x) {
    _in_bind_context = (x ? true : false);
  }

  return {
    yieldop       : yield_op,
    bind          : bind,
    new_branch    : new_branch,
    new_branch1   : new_branch1,
    new_handler   : new_handler,
    new_handler1  : new_handler1,
    handle_exn    : handle_exn,
    to_iter       : to_iter,
    yield_iter    : yield_iter,
    is_in_bind_context : is_in_bind_context,
    set_in_bind_context: set_in_bind_context,
  }
})();


// extend std_core with these primitives
$std_core = _export($std_core, {
  _yieldop      : Eff.yieldop,
  _bind         : Eff.bind,
  _new_branch   : Eff.new_branch,
  _new_branch1  : Eff.new_branch1,
  _new_handler  : Eff.new_handler,
  _new_handler1 : Eff.new_handler1,
  _handle_exn   : Eff.handle_exn,
  _is_in_bind_context : Eff.is_in_bind_context,
  _set_in_bind_context: Eff.set_in_bind_context,
});


