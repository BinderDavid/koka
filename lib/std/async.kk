/*---------------------------------------------------------------------------
  Copyright 2012-2016 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

// Asynchronous primitives
module std/async

import std/num/int32
import std/time/duration

extern include {
  js file "async-inline.js"
}

type extend exception-info {
  Cancel
}

public fun cancelable( action : (cancel:() -> exn ()) -> <async,exn|e> a ) : <async,exn|e> maybe<a> { 
  val caction = create-cancelable-prim(action)
  catch { Just(caction()) } fun(exn) {
    match(exn.info) {
      Cancel -> {
        check-canceled() // will re-throw if action is not the one that was canceled
        Nothing
      }
      _ -> throw(exn)
    }
  }
}

extern create-cancelable-prim( action : (cancel:() -> exn ()) -> <async,exn|e> a ) : (() -> <async,exn|e> a) {
  js "_create_cancelable"
}

extern check-canceled() : exn () {
  js "_check_canceled"
}

public extern unsafe-register-cancel( cb : a -> <async|e> (), cancel : () -> exn () ) : exn () {
  js "_register_cancel"
}

public fun timeout( secs : duration, action : () -> <async,exn|e> a ) : <async,exn|e> maybe<a> {
  cancelable fun(cancel) {
    val h = set-timeout( fun(_){ cancel(); }, secs.milli-seconds.int32 )
    on-exit({ cancel-timeout(h) }, action)
  }
}

// ----------------------------------------------------------------------------
// Asynchronous operations are supported through the `:async` effect.
// ----------------------------------------------------------------------------

public fun async-fork( action1 : () -> <async|e> a, action2 : () -> <async|e> ()) : <async|e> a {
  val xs = async-all( [{ Just(action1()) }, { action2(); Nothing }] )
  unsafe-noexn{ xs.head.unjust.unjust }
}

// Execute `actions` asynchronously, i.e. interleaved around asynchronous operations.
// When all operations have terminated (or raised an exception) return the list of
// results (or raise the first raised exception).
public fun async-all( actions : list<() -> <async|e> a> ) : <async|e> list<a> {
  async-all-prim( actions.vector ).list;
}


extern async-all-prim( actions : vector<() -> <async|e> a> ) : <async|e> vector<a>  {
  js inline "$asyncAll(actions,_k)"
}

extern async-callback0( action: () -> e (), haserr : bool, nodup : bool ) : (() -> e ()) {
  js "$asyncCallback"
}

public fun unsafe-async-yield-err( action : (cb : a -> <async|e> ()) -> <async|e> () ) : <async|e> a {
  unsafe-async-yield( action, True ) 
}

public extern unsafe-async-yield( action : (cb : a -> <async|e> ()) -> <async|e> (), haserr : bool = False ) : <async|e> a {
  js inline "(#1)( $asyncCallback(_k,#2), $std_core.async_yield)"
}

// Wait (asynchronously) for `secs` seconds. 
// Use `wait(0)` to yield to other asynchronous operations.
public fun wait( secs : double ) : <async,exn> () {
  wait(secs.duration)
}

// Wait (asynchronously) for `secs` seconds duration. 
// Use `wait(0)` to yield to other asynchronous operations.
public fun wait( secs : duration ) : <async,exn> () {
  unsafe-async-yield( fun(cb) { 
    val msecs = secs.milli-seconds.int32
    val h = set-timeout( cb, msecs)
    if (msecs > 100.int32) { 
      unsafe-register-cancel( cb, fun(){ cancel-timeout(h) } )
    }
  })
}

type timeout-handle

extern set-timeout( action : (()) -> e (), ms : int32 ) : exn timeout-handle {
  js "setTimeout"
}

extern cancel-timeout( h : timeout-handle ) : exn () {
  js "clearTimeout"
}


// Execute any asynchronous actions inside `action` synchronously, ie. 
// no other actions are interleaved.
public fun synchronous( action : () -> <async|e> a ) : e a {
  start-synchronous()
  on-exit(end-synchronous){ unsafe-total(action) }
}

extern start-synchronous() : () {
  js "$asyncStartSynchronous"
}

extern end-synchronous() : () {
  js "$asyncEndSynchronous"
}

