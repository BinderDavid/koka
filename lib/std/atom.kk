/*---------------------------------------------------------------------------
  Copyright 2012 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

/* This module provides a perfect hash for strings.

   These `:atom`s can be compared very efficiently. If you compare only
   seldomly, a straight string compare will probably be faster since 
   creating an atom takes some initial work.
*/
module std/atom

import std/vector

external include {
  cs file "atom-inline.cs"
  js file "atom-inline.js"
}

// An `atom` corresponding to the empty string
public val atom-null = atom("")

// An atom is an abstract representation of a string that can be compared quickly
abstract struct atom( value: int )  

// Create an `:atom`
extern atom-create : ( s : string ) -> int {
  cs "Atom.Create"
  js "$atomCreate"
}

// Convert an `:atom` back to a `:string`
extern atom-to-string : ( x : int ) -> string {
  cs "Atom.ToString"
  js "$atomToString"
}

// Create an `:atom`
public fun atom( s : string ) : atom {
  atom-create(s).Atom
}

// Convert an `:atom` back to a `:string`
public fun string( a : atom ) : string {
  atom-to-string(a.value)
}

// Compare two atoms for equality
public fun (==)(x : atom, y : atom) : bool  -> (x.value == y.value)

// Compare if two atoms are not equal.
public fun (!=)(x : atom, y : atom) : bool  -> (x.value != y.value)


extern atom-matcher : ( xs : vector<string>, ignoreCase: int ) -> any
{
  cs "Atom.CreateMatcher"
  js "$atomCreateMatcher"
}

extern atom-match : ( m : any, s : string, ignoreCase: int ) -> int
{
  cs "Atom.Match"
  js "$atomMatch"
}

// Create an efficient string matcher which given a list of strings returns a matching function.
public fun matcher( xs : list<string>, ignoreCase: bool = False ) : (( s : string ) -> bool)
{
  matcher(vector(xs), ignoreCase)
}

// Create an efficient string matcher which given a vector of strings returns a matching function.
public fun matcher( xs : vector<string>, ignoreCase: bool = False ) : (( s : string ) -> bool)
{
  val m = atom-matcher( xs, ignoreCase.int );
  return fun(s) { return bool(atom-match(m,s,int(ignoreCase))); }
}
