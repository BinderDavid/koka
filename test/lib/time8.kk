/*----------------------------------------------------------------------------
   Copyright (C) 2012-2016 Microsoft Corporation
    
   Licensed under the Apache License, Version 2.0 ("The Licence"). You may not
   use this file except in compliance with the License. A copy of the License
   can be found in the file "license.txt" at the root of this distribution.
----------------------------------------------------------------------------*/

// Test UT1 time
module time8

import std/os/path
import std/os/file
public import std/num/fixed
public import std/time
public import std/time/ut1

/*----------------------------------------------------------------------------
  Testing
----------------------------------------------------------------------------*/

fun check(name : string, res : string, tst : () -> io string ) : io () {
  val got = tst()
  println(name.pad-right(14,' ') + ": " 
    + (if (got == res) then "ok: " + res 
                       else "FAILED!:\n expect: " + res + "\n gotten: " + got + "\n"))
}


public val sofai = instant(2006,1,15,21,24,37,0.5)

public fun test-ut1() {
  println("load UT1 calendar")
  val calut1= on-exn(cal-ut1-approx){ cal-ut1-load(verbose=True) }
  check("sofa-ut1","2006-01-15T21:24:37.834075267Z UT1"){  sofai.time(cal=calut1).show  } // astropy: 37.834077793, sofa: 37.834100 ; non-interpolated: 37.834087400Z

  // across a leap second
  check("leap-ut1a","2012-06-30T23:59:57.413235998Z UT1"){ instant(2012,6,30,23,59,58).time(cal=calut1).show }
  check("leap-ut1b","2012-06-30T23:59:58.413235999Z UT1"){ instant(2012,6,30,23,59,59).time(cal=calut1).show }
  check("leap-ut1c","2012-06-30T23:59:59.413235999Z UT1"){ instant(2012,6,30,23,59,60).time(cal=calut1).show }
  check("leap-ut1d","2012-07-01T00:00:00.412236Z UT1")   { instant(2012,6,30,23,59,60,0.999).time(cal=calut1).show }
  check("leap-ut1e","2012-07-01T00:00:00.413236Z UT1")   { instant(2012,7,1,0,0,0).time(cal=calut1).show }
  check("leap-ut1f","2012-07-01T00:00:01.413236002Z UT1"){ instant(2012,7,1,0,0,1).time(cal=calut1).show }

}

public fun main() {
  test-ut1()
}