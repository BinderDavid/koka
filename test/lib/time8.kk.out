load UT1 calendar
sofa-ut1 : ok: 2006-01-15T21:24:37.834075267Z UT1
leap-ut1a : ok: 2012-06-30T23:59:57.413235998Z UT1
leap-ut1b : ok: 2012-06-30T23:59:58.413235999Z UT1
leap-ut1c : ok: 2012-06-30T23:59:59.413235999Z UT1
leap-ut1d : ok: 2012-07-01T00:00:00.412236Z UT1
leap-ut1e : ok: 2012-07-01T00:00:00.413236Z UT1
leap-ut1f : ok: 2012-07-01T00:00:01.413236002Z UT1